src/Data/Text/Display/Generic.hs

```diff
@@ -0,0 +1,127 @@
+--  Copyright   : © Hécate Moonlight, 2021
+--  License     : MIT
+--  Maintainer  : hecate@glitchbra.in
+--  Stability   : stable
+--
+--  Generic machinery for automatically deriving display instances for record types
+module Data.Text.Display.Generic where
+
+import Data.Text.Display.Core
+import Data.Text.Lazy.Builder (Builder)
+import qualified Data.Text.Lazy.Builder as TB
+import GHC.Generics
+import Generic.Data
+
+-- | Generic typeclass machinery for inducting on the structure
+-- of the type, such that we can thread `Display` instances through
+-- the structure of the type. The primary use case is for implementing
+-- `RecordInstance`, which does this "threading" for record fields. This
+-- machinery does, crucially, depend on child types (i.e. the type of a
+-- record field) having a `Display` instance.
+--
+-- @since 0.0.5.0
+class GDisplay1 f where
+  gdisplayBuilder1 :: f p -> Builder
+
+instance GDisplay1 V1 where
+  gdisplayBuilder1 x = case x of {}
+
+instance GDisplay1 U1 where
+  gdisplayBuilder1 _ = "()"
+
+-- | This is the most important instance, it can be considered as the "base case". It
+-- requires a non-generic `Display` instance. All this generic machinery can be conceptualized
+-- as distributing these `displayBuilder` across a product type.
+instance Display c => GDisplay1 (K1 i c) where
+  gdisplayBuilder1 (K1 a) = displayBuilder a
+
+instance (Constructor c, GDisplay1 f) => GDisplay1 (M1 C c f) where
+  gdisplayBuilder1 c@(M1 a)
+    | conIsRecord c = TB.fromString (conName c) <> "\n  { " <> gdisplayBuilder1 a <> "\n  }"
+    | conIsTuple c = TB.fromString (conName c) <> " ( " <> gdisplayBuilder1 a <> " )"
+    | otherwise = TB.fromString (conName c) <> " " <> gdisplayBuilder1 a
+    where
+      conIsTuple :: C1 c f p -> Bool
+      conIsTuple y =
+        tupleName (conName y)
+        where
+          tupleName ('(' : ',' : _) = True
+          tupleName _ = False
+
+instance (Selector s, GDisplay1 f) => GDisplay1 (M1 S s f) where
+  gdisplayBuilder1 s@(M1 a) =
+    if selName s == ""
+      then gdisplayBuilder1 a
+      else TB.fromString (selName s) <> " = " <> gdisplayBuilder1 a
+
+instance GDisplay1 f => GDisplay1 (M1 D s f) where
+  gdisplayBuilder1 (M1 a) = gdisplayBuilder1 a
+
+instance (GDisplay1 a, GDisplay1 b) => GDisplay1 (a :*: b) where
+  gdisplayBuilder1 (a :*: b) = gdisplayBuilder1 a <> "\n  , " <> gdisplayBuilder1 b
+
+instance (GDisplay1 a, GDisplay1 b) => GDisplay1 (a :+: b) where
+  gdisplayBuilder1 (L1 a) = gdisplayBuilder1 a
+  gdisplayBuilder1 (R1 b) = gdisplayBuilder1 b
+
+gdisplayBuilderDefault :: (Generic a, GDisplay1 (Rep a)) => a -> Builder
+gdisplayBuilderDefault = gdisplayBuilder1 . from
+
+-- | We leverage the `Generic.Data.GenericProduct` type to prevent consumers
+-- from deriving instances for sum types. Sum types should use a manual instance
+-- or derive one via `ShowInstance`.
+--
+-- @since 0.0.5.0
+instance (Generic a, GDisplay1 (Rep a)) => Display (GenericProduct a) where
```

Comment:
I don't know generic-data at all, what is the error message when you try to derive Display for a Sum?

Reply:
This is where it happens https://hackage.haskell.org/package/generic-data-1.1.0.0/docs/Generic-Data-Internal-Error.html#t:AssertNoSum. I agree that its not the best error message, but I was going to have to write some annoying type family machinery to get the appropriate custom error message, so I figured this was a fine interim solution.

Reply:
I see. This might seem petty but I'm not too chuffed about the fact that `generic-data` transitively depends on array, stm, StateVar and contravariant.

How annoying would be the tyfam solution? More annoying than what was in the pg-entity code?

Reply:
Let me see. I think it actually shouldn't be too bad.

Reply:
Why do you not want to derive `Display` for sums? The `GDisplay` instance for `(:+:)` is fine, isn't it? As far as I can tell, this really is a better fit for `Generically`, which is defined in `base` recently.

`GenericProduct` on its own doesn't do anything different from `Generically`. It was meant to address a strange corner case of generic deriving for `Monoid`. If you want to forbid generic deriving on sums, it doesn't matter whether the type is named `Generically` or `GenericProduct`, you have to add the relevant error message to the `(:+:)` instance.

Reply:
I'm open to Generically deriving the instances for Sums. I was just imagining that we wanted to constrain the way that some of the deriving via machinery was used. I actually just cribbed the AssertSum bit and placed it on the `RecordInstance` newtype.